crates/tx5/src/lib.rs

```diff
@@ -36,261 +36,177 @@ pub mod deps {
     pub use tx5_signal::deps::*;
 }
 
-use deps::{serde, serde_json};
-
-use tx5_core::Uniq;
 pub use tx5_core::{Error, ErrorExt, Id, Result, Tx5InitConfig, Tx5Url};
 
-pub mod actor;
-
-mod back_buf;
-pub use back_buf::*;
-
-/// Helper extension trait for `Box<dyn bytes::Buf + 'static + Send>`.
-pub trait BytesBufExt {
-    /// Convert into a `Vec<u8>`.
-    fn to_vec(self) -> Result<Vec<u8>>;
-}
-
-impl BytesBufExt for Box<dyn bytes::Buf + 'static + Send> {
-    fn to_vec(self) -> Result<Vec<u8>> {
-        use bytes::Buf;
-        use std::io::Read;
-        let mut out = Vec::with_capacity(self.remaining());
-        self.reader().read_to_end(&mut out)?;
-        Ok(out)
-    }
-}
-
-const FINISH: u64 = 1 << 63;
+mod ep3;
+pub use ep3::*;
 
-trait FinishExt: Sized {
-    fn set_finish(&self) -> Self;
-    fn unset_finish(&self) -> Self;
-    fn is_finish(&self) -> bool;
-}
+pub(crate) mod back_buf;
+pub(crate) use back_buf::*;
 
-impl FinishExt for u64 {
-    fn set_finish(&self) -> Self {
-        *self | FINISH
-    }
+pub(crate) mod proto;
 
-    fn unset_finish(&self) -> Self {
-        *self & !FINISH
-    }
+#[derive(Clone)]
+struct DoDrop(std::sync::Arc<tokio::task::JoinHandle<()>>);
 
-    fn is_finish(&self) -> bool {
-        *self & FINISH > 0
+impl Drop for DoDrop {
+    fn drop(&mut self) {
+        self.0.abort();
```

Comment:
Has tokio task cancellation always been a thing or is it a recent thing? I always remember tasks being uncancelable.

Reply:
[It's been around at least since tokio 1.0](https://docs.rs/tokio/1.0.1/tokio/task/struct.JoinHandle.html#method.abort)